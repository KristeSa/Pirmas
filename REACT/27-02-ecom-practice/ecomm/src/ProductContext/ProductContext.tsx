import { createContext } from "react";
import { TProductContext } from "./types";

const INITIAL_VALUE = {
  fetchedProducts: [],
  cartProducts: [],
  dispatch: () => {},
} as const;

export const ProductContext = createContext<TProductContext>(INITIAL_VALUE);
