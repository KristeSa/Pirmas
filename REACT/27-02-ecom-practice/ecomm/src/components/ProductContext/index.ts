export {ProductContext} from "./ProductContext"
export {productsReducer} from "./productsReducer"