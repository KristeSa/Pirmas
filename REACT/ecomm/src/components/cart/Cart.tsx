import { useReducer, useState } from "react";
import { cartReducer } from "./cartReducer";
import { CartItem, initialState, Reducer } from "./types";

export const Cart = () => {
  const [cartItem, setCartItem] = useState<CartItem[]>([]);
  const [state, dispatch] = useReducer<Reducer>(cartReducer, initialState);

  const removeItem = (item: CartItem) => {
    dispatch({ type: "REMOVE_ITEM", payload: item });
  };

  return (
    <>
      <h1>Shopping Cart</h1>
      <ul>
        {cartItem.map((item) => (
          <li key={item.id}>
            <img src={item.image} alt={item.title} />
            <span>{item.title}</span>
            <span>{item.price}</span>
            <button onClick={() => removeItem(item)}>Remove from Cart</button>
          </li>
        ))}
      </ul>
      <p>Total: {state.total}</p>
    </>
  );
};
