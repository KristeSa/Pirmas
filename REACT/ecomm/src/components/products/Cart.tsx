import { useContext } from "react"

export const ShoppingCart = () => {
    const { cart, removeItem } = useContext(CartContext);

    return (
        <>
          {todos.map((todo) => (
            <div key={todo.id}>
              <label>
                <input
                  type="checkbox"
                  checked={todo.complete}
                  onChange={() => handleComplete(todo)}
                />
                {todo.title}
              </label>
            </div>
          ))}
        </>
      );
    }
}