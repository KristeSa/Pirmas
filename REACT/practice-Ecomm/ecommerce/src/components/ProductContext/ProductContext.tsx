import { createContext, useState } from "react";

export const ProductsContext = createContext<any[]>([]);
const products = [
  { id: 1, name: "t-shirt", price: 19 },
  { id: 2, name: "t-shirt", price: 25 },
  { id: 3, name: "pants", price: 35 },
  { id: 4, name: "pants", price: 49 },
  { id: 5, name: "jeans", price: 79 },
  { id: 6, name: "jeans", price: 65 },
  { id: 7, name: "sweater", price: 40 },
];

export const CartProvider = ({ children }: any) => {
  const [items, setItems] = useState<any[]>([]);
  const addToCart = ({ name, price }: any) => {
    setItems((prevState) => [...prevState, { name, price }]);
  };

  return (
    <ProductsContext.Provider value={items}>
      {children}
    </ProductsContext.Provider>
  );
};
