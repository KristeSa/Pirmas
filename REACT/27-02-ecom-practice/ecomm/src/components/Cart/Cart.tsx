import { Box, Grid, Typography } from "@mui/material";
import { FC, useContext, useEffect, useState } from "react";
import { ProductContext } from "../ProductContext";
import { ProductActionButton } from "../Products/ProductActionButton";

export const Cart = () => {
  const { cartProducts } = useContext(ProductContext);
  const [isEmptyCart, setIsEmptyCart] = useState(true);

  const currencyFormat = new Intl.NumberFormat(navigator.language, {
    style: "currency",
    currency: "EUR",
  });

  const totalPrice = cartProducts.reduce(
    (curPrice, cartProduct) =>
      curPrice + (cartProduct.price || 0) * cartProduct.amount,
    0
  );

  useEffect(() => {
    setIsEmptyCart(cartProducts.length === 0);
  }, []);

  return (
    <>
      {isEmptyCart ? (
        <Typography
          component="h3"
          variant="h3"
          padding={2}
          fontSize="1.5rem"
          textAlign="center"
          borderTop="1px solid black"
          role="empty cart message"
        >
          Your cart is empty
        </Typography>
      ) : (
        <Box>
          <ul>
            {cartProducts.map((product) => (
              <li>
                <Grid
                  aria-label="cart-products"
                  item
                  xs={12}
                  container
                  padding={"10px"}
                  justifyContent={"space-between"}
                  bgcolor="cream"
                  border="1px solid black"
                  mx="auto"
                  sx={{ "& MuiTypography-root": { fontSize: "20px" } }}
                >
                  <Grid item xs={6}>
                    <Typography aria-label="product title">
                      {product.title}
                    </Typography>
                  </Grid>

                  <Grid item xs={3} display="flex" gap={"20px"}>
                    <ProductActionButton
                      color="success"
                      title="+"
                      type="addProduct"
                      aria-label="add button"
                      productId={product.id}
                    />

                    <Typography textAlign={"center"}>
                      {product.amount}
                    </Typography>

                    <ProductActionButton
                      color="error"
                      title="-"
                      type="removeProduct"
                      aria-label="remove button"
                      productId={product.id}
                    />
                  </Grid>

                  <Grid item xs={3}>
                    {product.price ? (
                      <Typography textAlign="right">
                        {currencyFormat.format(product.price)}
                      </Typography>
                    ) : null}
                  </Grid>
                </Grid>
              </li>
            ))}
          </ul>

          <Typography
            variant="h4"
            textAlign="right"
            mt="20px"
            aria-label="total price"
          >
            Total price: {currencyFormat.format(totalPrice)}
          </Typography>
        </Box>
      )}
    </>
  );
};
