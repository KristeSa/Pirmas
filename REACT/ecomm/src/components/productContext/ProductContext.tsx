import React, { createContext, Dispatch, FC, useReducer } from "react";
import { Actions, cartReducer, State } from "../products/cartReducer";
import { TProductContext } from "./types";

//const initialValue: TProductContext = { products: [], setProducts: () => {} };

const initialState = {
  counter: 0,
  products: [],
};

export const INITIAL_VALUE: TProductContext = {
  products: [],
  setProducts: () => {},
  cartProducts: [],
  setCartProducts: () => {},
};

export const ShoppingContext = createContext<{
  state: State;
  dispatch: Dispatch<Actions>;
}>({
  state: initialState,
  dispatch: () => null,
});

export const ShoppingProvider: FC = ({ children }) => {
  const [state, dispatch] = useReducer(cartReducer, initialState);

  return (
    <ShoppingContext.Provider value={{ state, dispatch }}>
      {children}
    </ShoppingContext.Provider>
  );
};

// const INITIAL_VALUE = {
//   products: [],
//   setProducts: () => {},
//   cartProducts: [],
//   setCartProducts: () => {},
// } satisfies TProductContext;

// const INITIAL_VALUE = {
//   products: [],
//   setProducts: () => {},
//   cartProducts: [],
//   setCartProducts: () => {},
// } as TProductContext;

export const ProductContext = createContext(INITIAL_VALUE);
