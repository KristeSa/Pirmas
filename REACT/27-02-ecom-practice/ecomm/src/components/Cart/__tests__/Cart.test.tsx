import { fireEvent, render, screen } from "@testing-library/react";
import { ProductContext } from "../../ProductContext";
import { Cart } from "../Cart";
import renderer from "react-test-renderer";

describe("Cart", () => {
  it("should render the cart with no items", () => {
    const emptyCart = 0;

    render(
      <ProductContext.Provider
        value={{
          cartProducts: [],
          fetchedProducts: [],
          dispatch: () => {},
        }}
      >
        <Cart />
      </ProductContext.Provider>
    );

    const productsInCart = expect(screen.getByRole("cart-products")).children
      .length;
    expect(productsInCart.length).toEqual(emptyCart);

    expect(screen.getByRole("empty cart message")).toBeVisible();
    expect(screen.getByRole("empty cart message").textContent).toBe(
      "Your cart is empty"
    );
  });

  it("renders cart products and total price", () => {
    const cartProducts = [
      {
        id: 1,
        title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
        price: 109.95,
        description:
          "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
        category: "men's clothing",
        image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
        rating: { rate: 3.9, count: 120 },
        amount: 1,
      },
      {
        id: 2,
        title: "Mens Casual Premium Slim Fit T-Shirts ",
        price: 22.3,
        description:
          "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
        category: "men's clothing",
        image:
          "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
        rating: { rate: 4.1, count: 259 },
        amount: 1,
      },
    ];

    render(
      <ProductContext.Provider
        value={{
          cartProducts,
          fetchedProducts: [],
          dispatch: () => {},
        }}
      >
        <Cart />
      </ProductContext.Provider>
    );

    expect(
      screen.getByText("Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops")
    ).toBeInTheDocument();
    expect(
      screen.getByText("Mens Casual Premium Slim Fit T-Shirts")
    ).toBeInTheDocument();
    expect(screen.getByText("Total price: €132.25")).toBeInTheDocument();
  });

  it("removes an item from the cart", () => {
    const cartProducts = [
      {
        id: 1,
        title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
        price: 109.95,
        description:
          "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
        category: "men's clothing",
        image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
        rating: { rate: 3.9, count: 120 },
        amount: 1,
      },
      {
        id: 2,
        title: "Mens Casual Premium Slim Fit T-Shirts ",
        price: 22.3,
        description:
          "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
        category: "men's clothing",
        image:
          "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
        rating: { rate: 4.1, count: 259 },
        amount: 1,
      },
    ];

    const removeFromCart = jest.fn();
    render(
      <ProductContext.Provider
        value={{
          cartProducts,
          fetchedProducts: [],
          dispatch: () => {},
        }}
      >
        <Cart />
      </ProductContext.Provider>
    );

    const removeButton = expect(screen.getByLabelText("remove button"));
    fireEvent.click(removeButton);
    expect(removeFromCart).toHaveBeenCalledWith(cartProducts[0]);
    expect(cartProducts.length).toBe(cartProducts.length - 1);
  });

  it("displays the total price of all items in the cart", () => {
    const cartProducts = [
      {
        id: 1,
        title: "Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops",
        price: 109.95,
        description:
          "Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday",
        category: "men's clothing",
        image: "https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg",
        rating: { rate: 3.9, count: 120 },
        amount: 1,
      },
      {
        id: 2,
        title: "Mens Casual Premium Slim Fit T-Shirts ",
        price: 22.3,
        description:
          "Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.",
        category: "men's clothing",
        image:
          "https://fakestoreapi.com/img/71-3HjGNDUL._AC_SY879._SX._UX._SY._UY_.jpg",
        rating: { rate: 4.1, count: 259 },
        amount: 1,
      },
    ];
    render(
      <ProductContext.Provider
        value={{
          cartProducts,
          fetchedProducts: [],
          dispatch: () => {},
        }}
      >
        <Cart />
      </ProductContext.Provider>
    );

    expect(screen.getByLabelText("total price")).toBeVisible();
    expect(screen.getByText("Total price: €132.25")).toBeInTheDocument();
  });

  it("matches snapshot", () => {
    const tree = renderer.create(<Cart />);

    expect(tree).toMatchSnapshot();
  });
});
