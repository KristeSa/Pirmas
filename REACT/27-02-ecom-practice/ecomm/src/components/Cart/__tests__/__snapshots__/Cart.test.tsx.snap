// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart matches snapshot 1`] = `
<h3
  className="MuiTypography-root MuiTypography-h3 css-1rov7ti-MuiTypography-root"
  role="empty cart message"
>
  Your cart is empty
</h3>
`;
